contracts/origination/OriginationControllerSTIRFRY.sol

```diff
@@ -0,0 +1,268 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.18;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+import "./OriginationControllerBase.sol";
+
+import "../interfaces/IOriginationControllerSTIRFRY.sol";
+import "../interfaces/IFeeController.sol";
+import "../interfaces/IVaultFactory.sol";
+
+import "../libraries/Constants.sol";
+
+import {
+    OCS_ZeroAddress,
+    OCS_InvalidStirfryPair,
+    OCS_InvalidPrincipalAmounts,
+    OCS_InvalidVaultAmount,
+    OCS_InvalidInterestAmounts
+} from "../errors/Lending.sol";
+
+/**
+ * @title OriginationControllerSTIRFRY
+ * @author Non-Fungible Technologies, Inc.
+ *
+ * STIRFRY - Set Term Interest Rate Fixed Rate Yield
+ *
+ * This contract is similar to the vanilla OriginationController, with a few key changes:
+ * 1). The collateral is collected from the lender instead of the borrower.
+ * 2). The borrower does not receive any principal at the start of the loan. Instead, the borrower deposits the
+ *     interest amount into the lenders vault. This vault is the vault that is being used for collateral in the loan.
+ * 3). There are no loan rollovers.
+ *
+ * In a STIRFRY loan scenario, a lender will mint a vault and deposit into the vault an ERC20 that accumulates some
+ * sort of variable yield. From the lender's perspective, they want to de-risk and lock in a fixed rate yield on
+ * their variable yield ERC20 token. This is where the borrower comes in, a borrower in a STIRFRY scenario will deposit
+ * the same ERC20 into the vault that the lender deposited. The amount of ERC20 that the borrower deposits is
+ * equivalent to the fixed rate yield that the lender is signaling in their signed loan terms. When a borrower
+ * deposits the fixed rate yield amount into the lender's vault, a loan will be originated. Upon loan origination,
+ * the borrower will not receive any principal. Instead, they are reserving the right to repay the loan at any time
```

Comment:
Repayment flows are all tested in RepaymentController.ts.

I tested one repayment flow in the STIRFRY tests, just to show what the e2e flow looks like and how the amount of tokens the borrower needs to front in  to receive the vaulted collateral. This can be extended to all repayment scenarios prior to the end of the loan. The borrower just needs to repay the prorata amount.